### PR TITLE
Template settings fixes/kira pilot

### LINKED ISSUE #3640: bug: template settings
[Screencast from 08-23-2022 02:37:39 PM.webm](https://user-images.githubusercontent.com/10648092/186152279-ba1e5055-64c5-4692-999f-ee47fa02e89a.webm)
There are multiple issues.

1. Does not allow saving if,

    - Description is longer than some characters. (Show a limit on the length of maximum characters) or show a message if the user exceeds that limit
    - if `MAX_TTL` time is 1 year = 31536000000 milliseconds.

2. Make `MAX_TTL` in hours

3. Allow setting `MAX_TTL` to `inf` or turning off the auto-shutdown for the template.

### PR DESCRIPTION
**Pull request by Kira-Pilot**
resolves #3640 from a FE perspective:

- TTL is now displayed/entered in hours
- TTL cannot be greater than 7 days (168 hours)
- TTL has validation error if above condition is not met
- TTL has helper text to add clarity
- Description has max char count of 128
- Description has validation error if above condition is not met
Screen Shot 2022-08-24 at 10 34 48 AM

Screen Shot 2022-08-24 at 11 12 34 AM

### CODE REVIEW

File: site/src/pages/TemplateSettingsPage/TemplateSettingsPage.test.tsx [hunk 1]
```diff
@@ -55,18 +62,73 @@ describe("TemplateSettingsPage", () => {
...
+  })
+
+  it("allows a ttl of 7 days", () => {
+    const values: UpdateTemplateMeta = {
+      ...validFormValues,
+      max_ttl_ms: 24 * 7,
+    }
+    const validate = () => validationSchema.validateSync(values)
+    expect(validate).not.toThrowError()
+  })
```
Comment (johnstcn) [inline site/src/pages/TemplateSettingsPage/TemplateSettingsPage.test.tsx:104]:
```suggestion
  it("allows a ttl of 7 days", () => {
    const values: UpdateTemplateMeta = {
      ...validFormValues,
      max_ttl_ms: 24 * 7,
    }
    const validate = () => validationSchema.validateSync(values)
    expect(validate).not.toThrowError()
  })

  it("allows max_ttl of 0", () => {
    const values: UpdateTemplateMeta = {
      ...validFormValues,
      max_ttl_ms: 0,
    }
    const validate = () => validationSchema.validateSync(values)
    expect(validate).not.toThrowError()
  })
```

**Review by johnstcn — APPROVED** [review]
Looking good! Backend ✅

File: site/src/components/WorkspaceScheduleForm/WorkspaceScheduleForm.tsx [hunk 2]
```diff
@@ -41,6 +41,7 @@ export const Language = {
   errorTime: "Time must be in HH:mm format (24 hours)",
   errorTimezone: "Invalid timezone",
   errorNoStop: "Time until shutdown must be greater than zero when auto-stop is enabled",
+  errorTtlMax: "Please enter a limit that is less than or equal to 168 hours (7 days).",
```
Comment (jsjoeio) [inline site/src/components/WorkspaceScheduleForm/WorkspaceScheduleForm.tsx:44]: nit: this line has punctuation but none of the others in the diff view do. doesn't matter which we use as long as we stay consistent IMO 
```suggestion
  errorTtlMax: "Please enter a limit that is less than or equal to 168 hours (7 days)",
```

File: site/src/pages/TemplateSettingsPage/TemplateSettingsForm.tsx [hunk 3]
```diff
@@ -20,16 +21,24 @@ export const Language = {
...
   maxTtlLabel: "Auto-stop limit",
   iconLabel: "Icon",
-  // This is the same from the CLI on https://github.com/coder/coder/blob/546157b63ef9204658acf58cb653aa9936b70c49/cli/templateedit.go#L59
-  maxTtlHelperText: "Edit the template maximum time before shutdown in seconds",
   formAriaLabel: "Template settings form",
   selectEmoji: "Select emoji",
+  ttlMaxError: "Please enter a limit that is less than or equal to 168 hours (7 days).",
+  descriptionMaxError: "Please enter a description that is less than or equal to 128 characters.",
+  ttlHelperText: (ttl: number): string =>
+    `Workspaces created from this template may not remain running longer than ${ttl} hours.`,
```
Comment (jsjoeio) [inline site/src/pages/TemplateSettingsPage/TemplateSettingsForm.tsx:29]: same comment as above 👆🏼 use punctuation or don't (your call) but hoping we can promote consistency

File: site/src/pages/TemplateSettingsPage/TemplateSettingsForm.tsx [hunk 4]
```diff
@@ -55,11 +64,18 @@ export const TemplateSettingsForm: FC<TemplateSettingsForm> = ({
     initialValues: {
       name: template.name,
       description: template.description,
-      max_ttl_ms: template.max_ttl_ms,
+      // on display, convert from ms => hours
+      max_ttl_ms: template.max_ttl_ms / 3600000,
```
Comment (jsjoeio) [inline site/src/pages/TemplateSettingsPage/TemplateSettingsForm.tsx:68]: nit: add comment explaining the 3600000 so future us remember what that represents

File: site/src/pages/TemplateSettingsPage/TemplateSettingsForm.tsx [hunk 5]
```diff
@@ -55,11 +64,18 @@ export const TemplateSettingsForm: FC<TemplateSettingsForm> = ({
...
       icon: template.icon,
     },
     validationSchema,
-    onSubmit,
+    onSubmit: (formData) => {
+      // on submit, convert from hours => ms
```
Comment (jsjoeio) [inline site/src/pages/TemplateSettingsPage/TemplateSettingsForm.tsx:73]: nice comment 👍🏼

File: site/src/pages/TemplateSettingsPage/TemplateSettingsForm.tsx [hunk 6]
```diff
@@ -148,19 +164,18 @@ export const TemplateSettingsForm: FC<TemplateSettingsForm> = ({
...
           variant="outlined"
-          // Display seconds from ms
-          value={form.values.max_ttl_ms ? form.values.max_ttl_ms / 1000 : ""}
-          // Convert ms to seconds
-          onChange={(event) =>
-            form.setFieldValue("max_ttl_ms", Number(event.currentTarget.value) * 1000)
-          }
+          type="number"
         />
+        {!form.errors.max_ttl_ms && form.values.max_ttl_ms && (
```
Comment (jsjoeio) [inline site/src/pages/TemplateSettingsPage/TemplateSettingsForm.tsx:174]: Took me a second to parse this: if no form errors for `max_ttl_ms` and there's a value for `max_ttl_ms` _then_ render this helper component. Got it 👍🏼

File: site/src/pages/TemplateSettingsPage/TemplateSettingsPage.test.tsx [hunk 7]
```diff
@@ -55,18 +62,82 @@ describe("TemplateSettingsPage", () => {
...
+
+  it("disallows a description of 128 + 1 chars", () => {
+    const values: UpdateTemplateMeta = {
+      ...validFormValues,
+      description:
+        "Nam quis nulla. Integer malesuada. In in enim a arcu imperdiet malesuada. Sed vel lectus. Donec odio urna, tempus molestie, port a",
+    }
+    const validate = () => validationSchema.validateSync(values)
+    expect(validate).toThrowError(FormLanguage.descriptionMaxError)
```
Comment (jsjoeio) [inline site/src/pages/TemplateSettingsPage/TemplateSettingsPage.test.tsx:141]: Really nice tests 👏🏼

**Review by jsjoeio — COMMENTED** [review]
Looks great 🚀

File: site/src/pages/TemplateSettingsPage/TemplateSettingsForm.tsx (hunk 6 above)
Comment (Kira-Pilot) [reply site/src/pages/TemplateSettingsPage/TemplateSettingsForm.tsx:174]: I'll add a comment!

**Pull request merged**